### ISSUE TITLE
Updating carbon via pip removes older data

### ISSUE THREAD

**Issue by redbaron4** [graphite-project/carbon#711]
I am not 100% sure this is a bug but I would appreciate if anyone could shed more light on this weird behaviour.

I had a running graphite-1.0.0 instance which is managed using pip-9.0.1

Today I ran a `pip install carbon==1.0.2` and it logged the following entry

```
Installing collected packages: carbon
  Found existing installation: carbon 1.0.0
    Uninstalling carbon-1.0.0:
DEPRECATION: --allow-all-external has been deprecated and will be removed in the future. Due to changes in the repository protocol, it no longer has any effect.
Exception:
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/pip-9.0.1-py2.7.egg/pip/basecommand.py", line 215, in main
    status = self.run(options, args)
  File "/usr/lib/python2.7/site-packages/pip-9.0.1-py2.7.egg/pip/commands/install.py", line 342, in run
    prefix=options.prefix_path,
  File "/usr/lib/python2.7/site-packages/pip-9.0.1-py2.7.egg/pip/req/req_set.py", line 778, in install
    requirement.uninstall(auto_confirm=True)
  File "/usr/lib/python2.7/site-packages/pip-9.0.1-py2.7.egg/pip/req/req_install.py", line 754, in uninstall
    paths_to_remove.remove(auto_confirm)
  File "/usr/lib/python2.7/site-packages/pip-9.0.1-py2.7.egg/pip/req/req_uninstall.py", line 115, in remove
    renames(path, new_path)
  File "/usr/lib/python2.7/site-packages/pip-9.0.1-py2.7.egg/pip/utils/__init__.py", line 267, in renames
    shutil.move(old, new)
  File "/usr/lib64/python2.7/shutil.py", line 299, in move
    rmtree(src)
  File "/usr/lib64/python2.7/shutil.py", line 247, in rmtree
    rmtree(fullname, ignore_errors, onerror)
  File "/usr/lib64/python2.7/shutil.py", line 256, in rmtree
    onerror(os.rmdir, path, sys.exc_info())
  File "/usr/lib64/python2.7/shutil.py", line 254, in rmtree
    os.rmdir(path)
OSError: [Errno 39] Directory not empty: '/opt/graphite/storage/whisper/collectd'

```

Managed to update to 1.0.2 after a couple of tries, but all my older metrics are no longer there.

I think this has something to do with the fact that storage dirs are a part of `data_files` directive and the fact that pip will remove `data_files` while updating the packaging.

**Comment by deniszh** [comment]
Similar report for graphite-web - https://github.com/graphite-project/graphite-web/issues/2193
Still need to investigate that.

**Comment by benjolitz** [comment]
Well, this just bit me and destroyed the data. All of it.

Thankfully it's not of high value, being home metrics and all that. But wow. As far as surprises go, this one got me. Changing my configuration to point to a different storage path.

**Comment by DanCech** [comment]
I haven't done any work on the packaging side of things yet, but I will dig into this asap

**Comment by deniszh** [comment]
`storage/whisper` directory is in the list of `data_files` for a long time already - https://github.com/graphite-project/carbon/blob/master/setup.py#L61
looks like something was changed in setuptools for processing existing files in subdirs.
Maybe we should change `storage/whisper` to e.g. `storage/whisper/dummy.txt`

**Comment by benjolitz** [comment]
That’s a good idea. It makes sense that pip/setuptools would remove package data directories and files. A reasonable expectation would be that pip would remove the file (`storage/whispher/dummy.txt`) but leave the directory intact if it wasn’t empty.

**Comment by redbaron4** [comment]
@deniszh This is a great idea. In the interim, removing the `../../storage` entries from the following files ensures the update does not remove the stored metrics

For carbon:

`/opt/graphite/lib/carbon-VERSION-pyVERSION.egg-info/installed-files.txt`

For graphite-web:

`/opt/graphite/webapp/graphite_web-VERSION-pyVERSION.egg-info/installed-files.txt`

Probably the next release can use the `dummy.txt` idea and users can ensure data safety by manually changing these files _before_ they upgrade to the next (corrected) version.

**Comment by deniszh** [comment]
@redbaron4 : 
Good idea, will do exactly that!

**Comment by redbaron4** [comment]
This was fixed in carbon-1.1.3
